test/integration/controlplane/transformation/kmsv2_transformation_test.go

```diff
@@ -847,6 +848,194 @@ resources:
 	}
 }
 
+// TestKMSv2FeatureFlag is an integration test between KubeAPI and ETCD
+// Concretely, this test verifies the following:
+// 1. When feature flag is not enabled, loading a encryptionConfig with KMSv2 should fail
+// 2. When feature flag is enabled, loading a encryptionConfig with KMSv2 should work
+// 3. When feature flag is disabled, loading a encryptionConfig with a non-v2 provider should work.
+// without performing a storage migration, decryption of existing data encrypted with v2 should fail for Get and List operations.
+// New data stored in etcd will no longer be encrypted using the external kms provider with v2 API.
+// 4. when feature flag is re-enabled, loading a encryptionConfig with the same KMSv2 plugin from 2 should work,
+// decryption of data encrypted with v2 should work
+func TestKMSv2FeatureFlag(t *testing.T) {
+	encryptionConfig := `
+kind: EncryptionConfiguration
+apiVersion: apiserver.config.k8s.io/v1
+resources:
+  - resources:
+    - secrets
+    providers:
+    - kms:
+       apiVersion: v2
+       name: kms-provider
+       endpoint: unix:///@kms-provider.sock
+`
+	// When feature flag is not enabled, loading a encryptionConfig with KMSv2 should fail
+	test, err := newTransformTest(t, encryptionConfig, false, "", nil)
+	if err == nil || !strings.Contains(err.Error(), "timed out waiting for the condition") {
+		t.Fatalf("when feature flag is not enabled, loading a encryptionConfig with KMSv2 should have failed with: timed out waiting for the condition, encryptionConfig:\n%s\n actual error: %v", encryptionConfig, err)
+	}
+
+	// When feature flag is enabled, loading a encryptionConfig with KMSv1 and v2 should work
+	defer featuregatetesting.SetFeatureGateDuringTest(t, utilfeature.DefaultFeatureGate, features.KMSv2, true)()
+
+	providerName := "kms-provider"
+	pluginMock := kmsv2mock.NewBase64Plugin(t, "@kms-provider.sock")
+
+	storageConfig := framework.SharedEtcd()
+
+	test, err = newTransformTest(t, encryptionConfig, false, "", storageConfig)
+	if err != nil {
+		t.Fatalf("failed to start KUBE API Server with encryptionConfig\n %s, error: %v", encryptionConfig, err)
+	}
+	restart := 0
+	defer func() {
+		if restart == 0 {
```

Comment:
We have to do this here because `test` gets recreated when `newTransformTest` is called.

Reply:
nvm I thought you meant `defer test.cleanUp()` which would fail. I've updated to
```go
defer func() {
   test.cleanUp()
}()
```
which should use the latest `test`
Thanks for the feedback!